File: tags/4.0.1/Java/OPSJLib/src/ops/MessageFilterSet.java
/**
*
* Copyright (C) 2006-2009 Anton Gravestam.
*
* This file is part of OPS (Open Publish Subscribe).
*
* OPS (Open Publish Subscribe) is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.

* OPS (Open Publish Subscribe) is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License
* along with OPS (Open Publish Subscribe).  If not, see <http://www.gnu.org/licenses/>.
*/

package ops;

import java.util.Vector;
import ops.protocol.OPSMessage;

/**
 *
 * @author angr
 */
public class MessageFilterSet implements MessageFilter
{
    Vector<MessageFilter> filters = new Vector();
    public synchronized void addFilter(MessageFilter filter)
    {
        filters.add(filter);
    }
    public synchronized void removeFilter(MessageFilter filter)
    {
        filters.remove(filter);
    }

    public synchronized boolean applyFilter(OPSMessage o)
    {
        for (MessageFilter messageFilter : filters)
        {
            if(!messageFilter.applyFilter(o))
            {
                return false;
            }
        }
        return true;
    }

    public boolean contains(Object o)
    {
        return filters.contains(o);
    }




}

File: C#/Ops/MessageFilterSet.cs
///////////////////////////////////////////////////////////
//  MessageFilterSet.cs
//  Implementation of the Class MessageFilterSet
//  Created on:      12-nov-2011 09:25:31
//  Author:
///////////////////////////////////////////////////////////

using System.Collections.Generic;       // Needed for the "List"
using System.Runtime.CompilerServices;  // Needed for the "MethodImpl" synchronization attribute

namespace Ops 
{
	public class MessageFilterSet : IMessageFilter 
    {
        internal List<IMessageFilter> filters = new List<IMessageFilter>();

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void AddFilter(IMessageFilter filter)
        {
            filters.Add(filter);
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void RemoveFilter(IMessageFilter filter)
        {
            filters.Remove(filter);
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public bool ApplyFilter(OPSMessage o)
        {
            foreach (IMessageFilter messageFilter in filters)
            {
                if (!messageFilter.ApplyFilter(o))
                {
                    return false;
                }
            }
            return true;
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public bool Contains(IMessageFilter messageFilter)
        {
            return filters.Contains(messageFilter);
        }

	}

}